pkg/endpoint/bpf.go

```diff
@@ -212,76 +218,79 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
+
+				// Only create a redirect if the proxy is NOT running in a sidecar container
+				// with an HTTP parser. If running in a sidecar container and the parser
+				// is HTTP, just allow traffic to the port at L4 by setting the proxy port
+				// to 0.
+				if !e.hasSidecarProxy || l4.L7Parser != policy.ParserTypeHTTP {
+
+					pp := e.newProxyPolicy(l4, v, dstPort)
+					proxyPort, err, finalizeFunc, revertFunc := e.proxy.CreateOrUpdateRedirect(e.aliveCtx, &pp, proxyID, e, proxyWaitGroup)
+					if err != nil {
+						// Skip redirects that can not be created or updated.  This
+						// can happen when a listener is missing, for example when
+						// restarting and k8s delivers the CNP before the related
+						// CEC.
+						// Policy is regenerated when listeners are added or removed
+						// to fix this condition when the listener is available.
+						e.getLogger().WithField(logfields.Listener, pp.GetListener()).WithError(err).Debug("Redirect rule with missing listener skipped, will be applied once the listener is available")
+						continue
+					}
+					redirectPort = proxyPort
+					finalizeList.Append(finalizeFunc)
+					revertStack.Push(revertFunc)
+				}
+				if e.realizedRedirects == nil {
+					e.realizedRedirects = make(map[string]uint16)
+				}
+				if _, found := e.realizedRedirects[proxyID]; !found {
+					revertStack.Push(func() error {
+						delete(e.realizedRedirects, proxyID)
+						return nil
+					})
+				}
+				e.realizedRedirects[proxyID] = redirectPort
+				desiredRedirects[proxyID] = true
+
+				// Update the endpoint API model to report that Cilium manages a
+				// redirect for that port.
+				statsKey := policy.ProxyStatsKey(l4.Ingress, string(l4.Protocol), dstPort, redirectPort)
+				e.proxyStatisticsMutex.Lock()
+				proxyStats := e.getProxyStatisticsLocked(statsKey, string(l4.L7Parser), dstPort, l4.Ingress)
+				proxyStats.AllocatedProxyPort = int64(redirectPort)
+				e.proxyStatisticsMutex.Unlock()
+
+				updatedStats = append(updatedStats, proxyStats)
 			}
-			e.realizedRedirects[proxyID] = redirectPort
-
-			desiredRedirects[proxyID] = true
-
-			// Update the endpoint API model to report that Cilium manages a
-			// redirect for that port.
-			statsKey := policy.ProxyStatsKey(l4.Ingress, string(l4.Protocol), dstPort, redirectPort)
-			e.proxyStatisticsMutex.Lock()
-			proxyStats := e.getProxyStatisticsLocked(statsKey, string(l4.L7Parser), dstPort, l4.Ingress)
-			proxyStats.AllocatedProxyPort = int64(redirectPort)
-			e.proxyStatisticsMutex.Unlock()
 
-			updatedStats = append(updatedStats, proxyStats)
-
-			if e.desiredPolicy == e.realizedPolicy {
-				// Any map updates when a new policy has not been calculated are taken care by incremental map updates.
-				return 0, false
-			}
-			return redirectPort, true
+			// Any map updates when a new policy has not been calculated are taken care
+			// by incremental map updates, so return 'false' in that case.
+			return e.desiredPolicy != e.realizedPolicy
```

Comment:
They are the same pointer in steady state, but during a policy reconfiguration they are different. The pattern is that `e.desiredPolicy` is set to a newly computed policy, and then when it has been fully realized, `e.updateRealizedState()` is called to update `e.realizedPolicy` to point to the same policy.

Reply:
`e.desiredPolicy == e.realizedPolicy` when endpoint regeneration is triggered and the current policy has already been computed for this endpoint. It seems harmless to add the redirect entries to the policy map also in this case (even though it is likely a no-op). I added a note on this to `runPrecompilationSteps()`: 
```
 	// Set the computed policy as the "incoming" policy. This can fail if
 	// the endpoint's security identity changed during or after policy calculation.
+	// Note that the incoming policy can be the same as the previous policy in cases
+	// where an unnecessary policy computation was skipped. In that case
+	// e.desiredPolicy == e.realizedPolicy also after this call.
+	// TODO: Figure out how much of the following logic would be safe to skip if there
+	// is no new incoming policy.
 	if err := e.setDesiredPolicy(policyResult); err != nil {
 		return false, err
 	}
```

Reply:
Turns out incremental updates will start being bypassed if a redirect in a `desiredPolicy` mapstate is updated when it is in fact the same as `realizedPolicy`, I think what happens is that the datapath bpf map update is skipped due to the entry already existing in the `realizedPolicy`. So I had to keep the pointer comparison logic here.

Reply:
I was looking through this code a bit a while back trying to figure out what situations the desiredPolicy would be recreated. At the time it appeared to me that having the two separate maps was not actually meaningful since there was never a case where `e.desiredPolicy != e.realizedPolicy`

Even in the `e.setDesiredPolicy(policyResult)` case, if I look at where the `endpointPolicy` that the desiredPolicy is being reassigned to comes from, its just the same reference of e.desiredPolicy:

```go
	result := &policyGenerateResult{
		selectorPolicy:   e.selectorPolicy,
		endpointPolicy:   e.desiredPolicy,
		identityRevision: e.identityRevision,
	}
```

It seemed like desiredPolicy is only ever assigned to a new map when being constructured for newEndpoint/restoreEndpoint.

Reply:
This is where it is updated once a new policy has been calculated:
```
	// Consume converts a SelectorPolicy in to an EndpointPolicy
	result.endpointPolicy = result.selectorPolicy.Consume(e)
```

Reply:
And then:
```
	// Set the revision of this endpoint to the current revision of the policy
	// repository.
	e.setNextPolicyRevision(res.policyRevision)
	e.selectorPolicy = res.selectorPolicy
	e.desiredPolicy = res.endpointPolicy
```

Reply:
ah I see now, distill allocates a new MapState which gets repopulated.